### DISCUSSION TITLE
Sequental Order & Fallback

### DISCUSSION THREAD

**Discussion by c-michail** [mokeyish/smartdns-rs#369] [General]
Hello,
  Is there a way to query with sequental order ?
  I mean to query from the first server only and fallback to the next server if the first server is offline.

Thank you

**Comment by mokeyish** [comment]
There is no such mechanism at present, but it should not be difficult to implement.

Currently the strategy is handled here.
https://github.com/mokeyish/smartdns-rs/blob/6c19d4adaf3404d5eb98f2d30176a37ef7bae5cb/src/dns_mw_ns.rs#L221

**Reply by c-michail** [reply]
Nice here's what i want to do:

1) Bootstrap server ONLY to get the IP of the DoH upstream server (IOW i don't want to use the bootstrap server for anything else)
2) Primary DoH upstream server for resolving all the queries
3) A fallback server in case the primary DoH upstream server goes down.

Also it should check with a timer if the primary Doh is up again and switch from fallback.

In case of a Sequental mode the fallbacks can be the 2nd, 3rd etc upstream servers.

Thank you

ps: i don't know rust in order to help you out with the code

**Comment by mokeyish** [comment]
I haven't implemented what you mentioned in the past two days, but there are relevant adjustments.

For the https://dns.google/dns-query upstream, its domain name resolves to two IPs(8.8.4.4, 8.8.8.8). The selection of these two IPs will be based on query statistics.

Here is the code:

https://github.com/mokeyish/smartdns-rs/blob/ddaf54e937cb0ea092322cd0c03f8d9aa8529abb/src/dns_client.rs#L584-L588


Does this meet your needs?

**Reply by c-michail** [reply]
Yep seems to be doing the same as "strict-order" of dnsmasq!

Here's my config:
server 1.1.1.1 -bootstrap-dns -exclude-default-group
server-quic mydnsserver.mydomain.com

0.87 also fixed the problem getting results from upstream bootstrap as default upstream. 


So the next step would be to add a fallback server in case there is a network error or current dns upstream is unreachable.
eg: (adding a fallback parameter i guess it's enough and also a timeout) 

server 1.1.1.1 -bootstrap-dns -exclude-default-group -fallback # it servers as bootstrap only and fallback too
server-quic mydnsserver.mydomain.com -timeout=1s # the only server should do the queries if it's reachable/online.

The logic should be i guess:
```
if "Query from upstream server(s) with timeout" = unreachable or timeout 
{
  if  Query from fallback = unreachable or timeout  then
  {
    return "fallback server failed"
  }
  else
  {
    return "Query success from fallback"
  }
}
```

timeout be defaulted around 2000 msecs i guess which is enough for slow networks too or maybe you should skip the "unreachable" and use the timeout value only.

So in this case you're relaxed in case the primary DNS Server(s) failure because you have a fallback to query from.
 
Thank you

ps: this can be optimized in the future release with a timer to check which servers are available and disable/enable them so avoiding querying from offline servers. (like speedtest option does)

You can have a status on the list of the servers (eg. IsAvailable: BOOL) and the timer will do a udp/tcp test to determinate if they are online or not every xx seconds.

**Comment by mokeyish** [comment]
It's not  "strict-order" of dnsmasq. The `servers.sort_unstable()` will be sorted using:

https://github.com/hickory-dns/hickory-dns/blob/558e68c876c9a52d2acd6af85980d3ad25b263c0/crates/resolver/src/name_server/name_server_stats.rs#L202

**Reply by c-michail** [reply]
If you implement a "strict-order" strategy is the same having fallback servers.

**Comment by mokeyish** [comment]
There is one thing you need to know here. The existing strategy is for the group. For the upstream of the domain name, it may return two IP addresses, but in fact, there can only be one in this group. As for which IP address it is, it is based on the query statistics above.

**Reply by mokeyish** [reply]
For the group, there are three strategies: FirstPing, FastestIp, fastestResponse.

**Reply by c-michail** [reply]
I have set in the config: 
speed-check-mode ping,tcp:443,tcp:80
response-mode first-ping

i think the above is the best to get the fastest ip.
